libyul/optimiser/SyntacticalEquality.cpp

```diff
@@ -29,48 +29,143 @@ using namespace std;
+{
+	return
+		compareVector<Identifier, &SyntacticallyEqual::expressionEqual>(_lhs.variableNames, _rhs.variableNames) &&
+		compareSharedPtr<Expression, &SyntacticallyEqual::operator()>(_lhs.value, _rhs.value)
+	;
+}
+
+bool SyntacticallyEqual::statementEqual(VariableDeclaration const& _lhs, VariableDeclaration const& _rhs)
+{
+	// first visit expression, then variable declarations
+	if (!compareSharedPtr<Expression, &SyntacticallyEqual::operator()>(_lhs.value, _rhs.value))
 		return false;
-	// TODO This somehow calls strcmp - WHERE?
-
-	// TODO This should be replaced by some kind of AST walker as soon as it gets
-	// more complex.
-	if (_e1.type() == typeid(FunctionalInstruction))
-	{
-		auto const& e1 = boost::get<FunctionalInstruction>(_e1);
-		auto const& e2 = boost::get<FunctionalInstruction>(_e2);
-		return
-			e1.instruction == e2.instruction &&
-			equalVector(e1.arguments, e2.arguments);
-	}
-	else if (_e1.type() == typeid(FunctionCall))
-	{
-		auto const& e1 = boost::get<FunctionCall>(_e1);
-		auto const& e2 = boost::get<FunctionCall>(_e2);
-		return
-			equal(e1.functionName, e2.functionName) &&
-			equalVector(e1.arguments, e2.arguments);
-	}
-	else if (_e1.type() == typeid(Identifier))
-		return boost::get<Identifier>(_e1).name == boost::get<Identifier>(_e2).name;
-	else if (_e1.type() == typeid(Literal))
-	{
-		auto const& e1 = boost::get<Literal>(_e1);
-		auto const& e2 = boost::get<Literal>(_e2);
-		return e1.kind == e2.kind && e1.value == e2.value && e1.type == e2.type;
-	}
-	else
-	{
-		assertThrow(false, OptimizerException, "Invalid expression");
-	}
-	return false;
-}
-
-bool SyntacticalEqualityChecker::equalVector(vector<Expression> const& _e1, vector<Expression> const& _e2)
-{
-	return _e1.size() == _e2.size() &&
-		std::equal(begin(_e1), end(_e1), begin(_e2), SyntacticalEqualityChecker::equal);
+	return compareVector<TypedName, &SyntacticallyEqual::visitDeclaration>(_lhs.variables, _rhs.variables);
+}
+
+bool SyntacticallyEqual::statementEqual(FunctionDefinition const& _lhs, FunctionDefinition const& _rhs)
+{
+	// first visit parameter declarations, then body
+	if (!compareVector<TypedName, &SyntacticallyEqual::visitDeclaration>(_lhs.parameters, _rhs.parameters))
+		return false;
+	if (!compareVector<TypedName, &SyntacticallyEqual::visitDeclaration>(_lhs.returnVariables, _rhs.returnVariables))
+		return false;
+	return statementEqual(_lhs.body, _rhs.body);
+}
 
+bool SyntacticallyEqual::statementEqual(If const& _lhs, If const& _rhs)
+{
+	return
+		compareSharedPtr<Expression, &SyntacticallyEqual::operator()>(_lhs.condition, _rhs.condition) &&
+		statementEqual(_lhs.body, _rhs.body)
+	;
+}
+
+bool SyntacticallyEqual::statementEqual(Switch const& _lhs, Switch const& _rhs)
+{
+	return
+		compareSharedPtr<Expression, &SyntacticallyEqual::operator()>(_lhs.expression, _rhs.expression) &&
+		compareVector<Case, &SyntacticallyEqual::switchCaseEqual>(_lhs.cases, _rhs.cases)
```

Comment:
Could visit cases in order of case value

Reply:
Would be easier after #5792 is fixed.

Reply:
Added this now in the latest commit, but it could be simplified if #5793 is merged first.

Reply:
Rebased on top of #5793 now.